src_firmware/src/storage.c

```diff
@@ -67,91 +86,233 @@ uint8_t get_template_size(void)
+	}
+	if (alloc_size == 0) {
+		return ERR_EMPTY_TEMPLATE_LIST;
+	}
+
+	int error;
+
 	for (int i = 0; i < alloc_size; i++) {
-		(void)nvs_read(&fs, i + TEMPLATES_START_ID,
+		error = nvs_read(&fs, i + TEMPLATES_START_ID,
 			       &(templates[i]), sizeof(struct Template));
+
+		if (error < 0) {
+			handle_errno_error(error, __LINE__);
+			return ERR_ERROR_CODE_FROM_ERRNO;
+		}
+		if (error == 0) {
+			return ERR_NOTHING_READ_DURING_NVS_READ;
+		}
 	}
+	return SUCCESS;
 }
 
-int get_template_by_name(char *name, struct Template *result)
-{ // TODO return error code, id as pointer
+int get_template_and_id_by_name(char *name, struct Template *result, uint8_t *out_id)
+{
+	if (!initialized) {
+		return ERR_NOT_INITIALISED;
+	}
+	if (alloc_size == 0) {
+		return ERR_EMPTY_TEMPLATE_LIST;
+	}
+
+	int error;
+
 	for (int i = 0; i < alloc_size; i++) {
-		(void)nvs_read(&fs, i + TEMPLATES_START_ID, result, sizeof(struct Template));
+		error = nvs_read(&fs, i + TEMPLATES_START_ID, result, sizeof(struct Template));
 		if (strcmp(name, result->name) == 0) {
-			return i;
+			if (error < 0) {
+				handle_errno_error(error, __LINE__);
+				return ERR_ERROR_CODE_FROM_ERRNO;
+			}
+			if (error == 0) {
+				return ERR_NOTHING_READ_DURING_NVS_READ;
+			}
+
+			*out_id = i;
+			return SUCCESS;
 		}
 	}
-	return -1;
+
+	return ERR_COULDNT_FIND_TEMPLATE;
 }
 
 int get_current_template(struct Template *result)
 {
+	if (!initialized) {
+		return ERR_NOT_INITIALISED;
+	}
 	//TODO - what if speed was set outside of template?
 	//(apply template -> set speed -> get current template)
-	// TODO - also pass properly results from get_template_by_name comment
-	return get_template_by_name(current_template, result);
+	uint8_t idx; // TODO - turn it to NULL?
+	// TODO - additional error check if we are outside of template.
+	return get_template_and_id_by_name(current_template, result, &idx);
 }
 
-void set_current_template(char *name)
+int set_current_template(char *name)
 {
+	if (!initialized) {
+		return ERR_NOT_INITIALISED;
+	}
+	int error;
+
 	strcpy(current_template, name);
-	nvs_write(&fs, CURRENT_TEMPLATE_ID, current_template,
+	// TODO - also check, if there is any current template applied
+	// if not, there is nothing to set, so return error
+	error = nvs_write(&fs, CURRENT_TEMPLATE_ID, current_template,
 		  CONFIG_TEMPLATE_NAME_SIZE * sizeof(char));
+
+	if (error < 0) {
+		handle_errno_error(error, __LINE__);
+		return ERR_ERROR_CODE_FROM_ERRNO;
+	}
+	if (error == 0) {
+		return ERR_NOTHING_WRITTEN_DURING_NVS_WRITE;
+	}
+
+	return SUCCESS;
 }
 
-void remove_template_by_name(char *name)
+int remove_template_by_name(char *name)
 {
-	struct Template template_to_remove;
-	int idx;
-
+	if (!initialized) {
+		return ERR_NOT_INITIALISED;
+	}
 	if (alloc_size == 0) {
-		return;
+		return ERR_EMPTY_TEMPLATE_LIST;
 	}
-	idx = get_template_by_name(name, &template_to_remove);
-	if (idx >= 0) {
+
+	struct Template template_to_remove;
+	uint8_t idx;
+	int error;
```

Comment:
```suggestion
	error_codes_t error;
```

Declarations should be moved to the top of the scope as well

Reply:
errno - int stays.
But moved to the top